### ISSUE TITLE
Type annotations

### ISSUE THREAD

**Issue by MSeal** [jupyter/nbclient#82]
We're on python 3 only, let's make use of type annotations on our functions.

**Comment by davidbrochart** [comment]
Should we use stubs files or add type annotations in the source code?
Should we use python>=3.6 type annotation syntax?

**Comment by MSeal** [comment]
> Should we use stubs files or add type annotations in the source code?

I haven't used stubs much, just type annotations. I'm not sure how much benefit we'd get for using stubs? If you have more experience on that would love an opinion.

> Should we use python>=3.6 type annotation syntax?

Yep

**Comment by choldgraf** [comment]
Wait you mean we shouldn't interweave traitlets everywhere?

**Comment by choldgraf** [comment]
(that is a joke, I think it's a good idea)

**Comment by davidbrochart** [comment]
> I haven't used stubs much, just type annotations. I'm not sure how much benefit we'd get for using stubs? If you have more experience on that would love an opinion.

I don't have any experience in type annotation, but I'm thinking that the Jupyter projects will need to exchange type definition, right? Like nbclient will probably need a definition of a `KernelManager` and a `KernelClient` I guess.

**Comment by MSeal** [comment]
> I don't have any experience in type annotation, but I'm thinking that the Jupyter projects will need to exchange type definition, right? Like nbclient will probably need a definition of a KernelManager and a KernelClient I guess.

Doesn't need a type stub -- you can just import the class and use it in the annotation inline. The inline annotations actually read better for expectations.

For example:
```
from jupyter_client.managerabc import KernelManagerABC
from nbformat import NotebookNode
...

def __init__(self, nb: NotebookNode, km: KernelManagerABC=None, **kw):
  ...
```
works just fine

**Comment by MSeal** [comment]
> Wait you mean we shouldn't interweave traitlets everywhere?

I actually don't know how traitlets play with type annotations... that would be something to figure out I guess (joke aside)

**Comment by choldgraf** [comment]
my (poor) understanding of traitlets is that one of the reasons it was developed was to allow you to type inputs, so I think (?) to some degree they replicate functionality

**Comment by davidbrochart** [comment]
Yes but traitlets validate type at runtime, it's not static type analysis.

**Comment by MSeal** [comment]
They also provide a config interface for assignment from file / cml args.

**Comment by choldgraf** [comment]
ok ok I apologize for throwing shade at traitlets :-)

**Comment by MSeal** [comment]
lol, I don't particularly like the traitlet model either as it uses less than ideal software practices to achieve (high complexity in tracing, 10+ class inheritance chains, makes __init__ operations more complex, very obscure error message patterns when it fails, unexpected config inheritance patterns, etc) -- which I've been vocal about. I'd prefer removing traitlets from nbclient if we could.

I've left them when porting over because the file based config is used very very heavily in nbconvert and the users who use it like it a lot. So removing it from nbclient would require mapping that in nbconvert in the preprocessor layer... might be worth discussing doing before 1.0 of nbclient and 6.0 of nbconvert fully ship? Would be tough to choose.

**Comment by davidbrochart** [comment]
I guess if we just wanted the CLI part of traitlets, we could use [click](https://click.palletsprojects.com). But there's also the reactive part in traitlets (observing changes).

**Comment by choldgraf** [comment]
(I'm a big fan of click)

**Comment by davidbrochart** [comment]
I have started to work on it in #83.

**Comment by davidbrochart** [comment]
Fixed in #83
